core/src/main/scala/kafka/coordinator/transaction/TransactionMetadata.scala

```diff
@@ -380,50 +417,39 @@ private[transaction] class TransactionMetadata(val transactionalId: String,
     }
   }
 
-  private def prepareTransitionTo(updatedState: TransactionState,
-                                  updatedProducerId: Long,
-                                  updatedEpoch: Short,
-                                  updatedLastEpoch: Short,
-                                  updatedTxnTimeoutMs: Int,
-                                  updatedTopicPartitions: immutable.Set[TopicPartition],
-                                  updatedTxnStartTimestamp: Long,
-                                  updateTimestamp: Long): TxnTransitMetadata = {
-    prepareTransitionTo(updatedState, updatedProducerId, RecordBatch.NO_PRODUCER_ID, updatedEpoch, updatedLastEpoch, updatedTxnTimeoutMs, updatedTopicPartitions, updatedTxnStartTimestamp, updateTimestamp, TransactionVersion.TV_0)
-  }
-
-  private def prepareTransitionTo(updatedState: TransactionState,
-                                  updatedProducerId: Long,
-                                  nextProducerId: Long,
-                                  updatedEpoch: Short,
-                                  updatedLastEpoch: Short,
-                                  updatedTxnTimeoutMs: Int,
-                                  updatedTopicPartitions: immutable.Set[TopicPartition],
-                                  updatedTxnStartTimestamp: Long,
-                                  updateTimestamp: Long,
-                                  clientTransactionVersion: TransactionVersion): TxnTransitMetadata = {
+  private def prepareTransitionTo(state: TransactionState,
+                                  producerId: Long = this.producerId,
+                                  nextProducerId: Long = this.nextProducerId,
+                                  producerEpoch: Short = this.producerEpoch,
+                                  lastProducerEpoch: Short = this.lastProducerEpoch,
+                                  txnTimeoutMs: Int = this.txnTimeoutMs,
+                                  topicPartitions: immutable.Set[TopicPartition] = this.topicPartitions.toSet,
+                                  txnStartTimestamp: Long = this.txnStartTimestamp,
+                                  txnLastUpdateTimestamp: Long = this.txnLastUpdateTimestamp,
+                                  clientTransactionVersion: TransactionVersion = this.clientTransactionVersion): TxnTransitMetadata = {
```

Comment:
Is the idea to keep clientTransactionVersion the same until each PrepareCommit/PrepareAbortCall? Or will KIP-939 introduce changes to set this in another transactional state.

Reply:
Good question.  I don't think KIP-939 would rely on persisted value of client version, we should be able to just look at the values of the fields (nextProducerId, nextProducerEpoch) and figure out what logic needs to be done.

But it's a good point that this change actually makes the clientVersion "sticky" in the sense that once it's set it'll keep the value during the transitions.  Maybe to keep this change completely mechanical, transition to ongoing or empty should reset the version to TV_0.

Reply:
Yeah, Things may also change with the change to add partitions implicitly (as we will have the client version there)